File: tile-client/src/main/webapp/js/annotation/AnnotationUtil.js
/*
 * Copyright (c) 2014 Oculus Info Inc.
 * http://www.oculusinfo.com/
 *
 * Released under the MIT License.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */


define(function (require) {
    "use strict";



    var ANNOTATION_CLASS = "annotation",
        ANNOTATION_ATTRIBUTE_CLASS = ANNOTATION_CLASS + " annotation-attribute",
        ANNOTATION_INPUT_CLASS = ANNOTATION_CLASS + " annotation-input",
        ANNOTATION_LABEL_CLASS = ANNOTATION_CLASS + " annotation-label",
        ANNOTATION_TEXTAREA_CLASS = ANNOTATION_CLASS + " annotation-textarea",
        ANNOTATION_TEXT_DIV_CLASS = ANNOTATION_CLASS + " annotation-text-div",
        ANNOTATION_CONTENT_CLASS = ANNOTATION_CLASS + " annotation-content-container",
        ANNOTATION_BUTTON_CLASS = ANNOTATION_CLASS + " annotation-button",
        ANNOTATION_LEFT_BUTTON_CLASS = ANNOTATION_BUTTON_CLASS + " annotation-button-left",
        ANNOTATION_RIGHT_BUTTON_CLASS = ANNOTATION_BUTTON_CLASS + " annotation-button-right",
        ANNOTATION_POPUP_ID = "annotation-popup-id",
        ANNOTATION_CANCEL_BUTTON_ID = "annotation-popup-cancel",
        ANNOTATION_SAVE_BUTTON_ID = "annotation-popup-save",
        ANNOTATION_EDIT_BUTTON_ID = "annotation-popup-edit",
        ANNOTATION_REMOVE_BUTTON_ID = "annotation-popup-remove",
        ANNOTATION_POPUP_TITLE_ID = "annotation-popup-title-id",
        ANNOTATION_POPUP_GROUP_ID = "annotation-popup-group-id",
        ANNOTATION_POPUP_DESCRIPTION_ID = "annotation-popup-description-id",
        ANNOTATION_ACCORDION_ID = "annotation-accordion-id";

    return {

        createEditablePopup: function( groups, closeFunc, saveFunc, removeFunc ) {

            // edit popup is only possible on single features, so take only data entry
            var that = this,
                hasRemoveFunc = ( removeFunc !== undefined ),
                html = this.getEditablePopupHTML( this.getDataArray()[0], groups, hasRemoveFunc );

            // create popup with close callback
            this.createPopup( html, function() {
                closeFunc( that );
                return false;   // stop event propagation
            });

            // set save callback
            $( "#"+ANNOTATION_SAVE_BUTTON_ID ).click( function() {
                saveFunc( that );
                return false;   // stop event propagation
            });

            // set cancel callback
            $( "#"+ANNOTATION_CANCEL_BUTTON_ID ).click( function() {
                closeFunc( that );
                return false;   // stop event propagation
            });

            if ( hasRemoveFunc ) {
                // set remove callback
                $( "#"+ANNOTATION_REMOVE_BUTTON_ID ).click( function() {
                    removeFunc( that );
                    return false;   // stop event propagation
                });
            }
        },


        createDisplayPopup: function( bin, closeFunc, editFunc ) {

            var that = this,
                hasEditFunc = editFunc !== null,
                html;

            if ( bin.length > 1 ) {
                // aggregate
                html = this.getAggregateDisplayPopupHTML();

            } else {
                // single
                html = this.getSingleDisplayPopupHTML( this.getDataArray()[0], hasEditFunc );

            }

            // create popup with close callback
            this.createPopup( html, function() {
                //closeFunc( that );
                return false;   // stop event propagation
            });


            if ( bin.length > 1 ) {

                // set accordion
                $( "#"+ANNOTATION_ACCORDION_ID ).accordion({
                    active: false,
                    collapsible: true,
                    heightStyle: "content"
                });

            } else {

                if ( hasEditFunc ) {
                    // set save callback
                    $( "#"+ANNOTATION_EDIT_BUTTON_ID ).click( function() {
                        editFunc( that );
                        return false;   // stop event propagation
                    });
                }

            }

            // set cancel callback
            $( "#"+ANNOTATION_CANCEL_BUTTON_ID ).click( function() {
                closeFunc( that );
                return false;   // stop event propagation
            });

        },


        checkInputData: function() {

            var $title = $('#'+ANNOTATION_POPUP_TITLE_ID),
                $group = $('#'+ANNOTATION_POPUP_GROUP_ID),
                $description = $('#'+ANNOTATION_POPUP_DESCRIPTION_ID),
                INVALID_COLOR = {color: '#7e0004'},
                INVALID_COLOR_EFFECT_LENGTH_MS = 3000;

            // if entry is invalid, flash
            if ( $title.val() === "" ) {
                $title.effect("highlight", INVALID_COLOR, INVALID_COLOR_EFFECT_LENGTH_MS);
            }
            if ( $group.val() === "" ) {
                $group.effect("highlight", INVALID_COLOR, INVALID_COLOR_EFFECT_LENGTH_MS);
            }
            if ( $description.val() === "" ) {
                $description.effect("highlight", INVALID_COLOR, INVALID_COLOR_EFFECT_LENGTH_MS);
            }

            // check input values
            return ( $title.val() !== "" &&
                     $group.val() !== "" &&
                     $description.val() !== "" );
        },


        setDataFromPopup: function( annotation ) {

            var $title = $('#'+ANNOTATION_POPUP_TITLE_ID),
                $group = $('#'+ANNOTATION_POPUP_GROUP_ID),
                $description = $('#'+ANNOTATION_POPUP_DESCRIPTION_ID);

            annotation.group = $group.val();
            annotation.data.title = $title.val();
            annotation.data.comment = $description.val();
        },


        checkAndSetData: function() {

            var annotation = this.getDataArray()[0];

            if ( this.checkInputData() ) {
                this.setDataFromPopup( annotation );
                return true;
            }
            return false;
        },


        getSingleDisplayPopupHTML: function( annotation, isEditable ) {

            var html = "<div style='overflow:hidden'>";

            if ( isEditable ) {
                html += "<input id='"+ ANNOTATION_EDIT_BUTTON_ID + "' type='image' src='./images/edit-icon.png' width='17' height='17'>";
            }

            html += "<div id='" +ANNOTATION_POPUP_ID+ "' class='ui-widget-content'>"+
                        "<div class='"+ANNOTATION_CONTENT_CLASS+"'>"+
                            "<div class='"+ANNOTATION_ATTRIBUTE_CLASS+"'>" +
                                annotation.data.title +
                            "</div>"+
                            "<div class='"+ANNOTATION_ATTRIBUTE_CLASS+"'>" +
                                "Priority: "+ annotation.group +
                            "</div>"+
                            "<div class='"+ANNOTATION_TEXT_DIV_CLASS+"'>"+
                                annotation.data.comment +
                            "</div>"+
                        "</div>"+
                    "</div>"+
                "</div>";

            return html;
        },


        getAggregateDisplayPopupHTML: function( bin ) {

            var annotations = this.getDataArray(),
                i,
                html = "<div style='overflow:hidden'>"+
                            "<div id='" +ANNOTATION_POPUP_ID+ "' class='ui-widget-content'>"+
                                "<div class='"+ANNOTATION_CONTENT_CLASS+"'>"+
                                    "<div id='"+ANNOTATION_ACCORDION_ID+"'>";

            for (i=0; i<bin.length; i++) {
                html += "<h3>" + bin.title + "</h3>"+
                            "<div>"+
                                "<div class='"+ANNOTATION_ATTRIBUTE_CLASS+"'>" +
                                    "Priority: "+ annotations[i].group +
                                "</div>"+
                                "<div class='"+ANNOTATION_TEXT_DIV_CLASS+"'>"+
                                    annotations[i].data.comment +
                                "</div>"+
                            "</div>";
            }

            return html + "</div></div></div></div>";
        },


        getEditablePopupHTML: function( annotation, groups, includeRemoveButton ) {

            var TITLE_PLACEHOLDER = ' Enter title',
                PRIORITY_PLACEHOLDER = 'Enter group',
                DESCRIPTION_PLACEHOLDER = ' Enter description',
                titleVal= annotation.data.title || "",
                descriptionVal = annotation.data.comment || "",
                removeButtonHTML = includeRemoveButton ? "<button class='"+ANNOTATION_LEFT_BUTTON_CLASS+"' id='" + ANNOTATION_REMOVE_BUTTON_ID + "'>Remove</button>" : "";

                function getSelectHTML() {

                    var html = "<select class='"+ANNOTATION_INPUT_CLASS+"' id='"+ ANNOTATION_POPUP_GROUP_ID+"'>",
                        i;

                    for (i=0; i<groups.length; i++) {
                        if ( groups[i] === annotation.group ) {
                            html += "<option selected='true' value='" +groups[i] +"'>"+groups[i]+"</option>";
                        } else {
                            html += "<option value='" +groups[i] +"'>"+groups[i]+"</option>";
                        }
                    }

                    if ( annotation.group === undefined ) {
                        html += "<option selected='true' style='display:none;' value=''>"+PRIORITY_PLACEHOLDER+"</option>";
                    }

                    return html + "</select>";
                }

                return  "<div style='overflow:hidden'>"+
                            "<div id='" +ANNOTATION_POPUP_ID+ "' class='ui-widget-content'>"+
                            "<div class='"+ANNOTATION_CONTENT_CLASS+"'>"+
                                "<div class='"+ANNOTATION_ATTRIBUTE_CLASS+"'>" +

                                    "<label class='"+ANNOTATION_LABEL_CLASS+"'>Title: </label>" +
                                    "<input class='"+ANNOTATION_INPUT_CLASS+"' id='"+ ANNOTATION_POPUP_TITLE_ID +"' type='text' placeholder='"+TITLE_PLACEHOLDER+"' value='"+ titleVal +"'>" +

                                "</div>"+
                                "<div class='"+ANNOTATION_ATTRIBUTE_CLASS+"'>" +

                                    "<label class='"+ANNOTATION_LABEL_CLASS+"'>Priority: </label>" +
                                    getSelectHTML()+

                                "</div>"+
                                "<div class='"+ANNOTATION_ATTRIBUTE_CLASS+"'> Description: </div>" +
                                "<textarea class='"+ANNOTATION_TEXTAREA_CLASS+"' id='"+ ANNOTATION_POPUP_DESCRIPTION_ID +"' placeholder='"+DESCRIPTION_PLACEHOLDER+"'>"+ descriptionVal +"</textarea> "+
                            "</div>"+

                        "</div>"+
                        removeButtonHTML +
                        "<button class='"+ANNOTATION_RIGHT_BUTTON_CLASS+"' id='" + ANNOTATION_CANCEL_BUTTON_ID + "'>Cancel</button>"+
                        "<button class='"+ANNOTATION_RIGHT_BUTTON_CLASS+"' id='" + ANNOTATION_SAVE_BUTTON_ID + "'>Save</button>"+
                    "</div>";
        }

    };


});

File: annotation-service/src/main/java/com/oculusinfo/annotation/util/AnnotationUtil.java
/*
 * Copyright (c) 2014 Oculus Info Inc. http://www.oculusinfo.com/
 * 
 * Released under the MIT License.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.oculusinfo.annotation.util;


import com.oculusinfo.annotation.AnnotationBin;
import com.oculusinfo.annotation.AnnotationData;
import com.oculusinfo.annotation.AnnotationTile;
import com.oculusinfo.annotation.index.AnnotationIndexer;
import com.oculusinfo.binning.BinIndex;
import com.oculusinfo.binning.TileIndex;
import com.oculusinfo.binning.impl.DenseTileData;
import com.oculusinfo.factory.util.Pair;
import org.json.JSONArray;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;


public class AnnotationUtil {

    private static final Logger LOGGER = LoggerFactory.getLogger( AnnotationUtil.class );

	public static void printData( Map<BinIndex, List<AnnotationData<?>>> dataMap ) {
		for ( List<AnnotationData<?>> annotations : dataMap.values() ) {
			printData( annotations );
		}
	}

	public static void printData( List<AnnotationData<?>> annotations ) {
		
		for ( AnnotationData<?> annotation : annotations ) {				
			try {
				LOGGER.debug( annotation.toJSON().toString( 4 ) );
			} catch ( Exception e ) {
				e.printStackTrace();
			}	
		}
	}

	public static void printTiles( List<AnnotationTile> tiles ) {
		
		for ( AnnotationTile tile : tiles ) {
			try {
				LOGGER.debug( tileToJSON( tile ).toString( 4 ) );
			} catch ( Exception e ) { e.printStackTrace(); }
					
		}
	}
	
	
	public static boolean compareData( List<AnnotationData<?>> as, List<AnnotationData<?>> bs ) {
		
		for ( AnnotationData<?> a : as ) {
			int foundCount = 0;
			for ( AnnotationData<?> b : bs ) {				
				if ( compareData(a, b ) ) {
					foundCount++;
				}
			}	
			if ( foundCount != 1 ) {
				LOGGER.debug( "Data lists do not match" );
				return false;
			}
		}
		return true;
	}
	
	public static boolean compareTiles( List<AnnotationTile> as, List< AnnotationTile > bs ) {
		
		for ( AnnotationTile a : as ) {
			int foundCount = 0;
			for ( AnnotationTile b : bs ) {
				if ( compareTiles( a, b, false ) ) {
					foundCount++;
				}
			}	
			if ( foundCount != 1 ) {
				LOGGER.debug( "Tile lists do not match" );
				return false;
			}
		}
		return true;
	}

    public static <T> boolean compareAttributes( T a, T b ) {

        if ( a == null || b == null ) {
            return ( a == b );
        } else {
            return a.equals( b );
        }
    }
	
	
	public static boolean compareData( AnnotationData<?> a, AnnotationData<?> b ) {
		
		if ( !a.getUUID().equals( b.getUUID() ) ) {
			LOGGER.debug( "UUID are not equal" );
			return false;
		}

        if ( !compareAttributes( a.getX(), b.getX() ) ) {
            LOGGER.debug( "X values are not equal" );
            return false;
		}

        if ( !compareAttributes( a.getY(), b.getY() ) ) {
            LOGGER.debug( "Y values are not equal" );
            return false;
        }

        if ( !compareAttributes( a.getX0(), b.getX0() ) ) {
            LOGGER.debug( "X0 values are not equal" );
            return false;
        }

        if ( !compareAttributes( a.getY0(), b.getY0() ) ) {
            LOGGER.debug( "Y0 values are not equal" );
            return false;
        }

        if ( !compareAttributes( a.getX1(), b.getX1() ) ) {
            LOGGER.debug( "X1 values are not equal" );
            return false;
        }

        if ( !compareAttributes( a.getY1(), b.getY1() ) ) {
            LOGGER.debug( "Y1 values are not equal" );
            return false;
        }

		if ( !a.getLevel().equals( b.getLevel() ) ) {
			LOGGER.debug( "Level values are not equal" );
			return false;
		}
		
		if ( !a.getData().toString().equals( b.getData().toString() ) ) {
			LOGGER.debug( "Data objects are not equal" );
			return false;
		}		
		return true;
	}
	
	
	public static boolean compareTiles( AnnotationTile a, AnnotationTile b, boolean verbose ) {
		
		List<Pair<String, Long>> aReferences = a.getAllCertificates();
		List<Pair<String, Long>> bReferences = b.getAllCertificates();
		
		if ( !a.getDefinition().equals( b.getDefinition() ) ) {
			LOGGER.debug( "Bin indices are not equal");
			return false;
		}
		
		if ( aReferences.size() != bReferences.size() ) {
			LOGGER.debug( "Reference counts are not equal");
			return false;		
		}
			
		for ( Pair<String, Long> aRef : aReferences ) {
			int foundCount = 0;
			for ( Pair<String, Long> bRef : bReferences ) {
				if ( aRef.equals( bRef ) ) {
					foundCount++;
				}
			}		
			if ( foundCount != 1 ) {
				LOGGER.debug( "Reference lists are not equal");
				return false;
			}			
		}

		return true;
	}


	public static List<TileIndex> tilesToIndices( List< AnnotationTile > tiles ) {
		List<TileIndex> indices = new ArrayList<>();
		for ( AnnotationTile tile : tiles ) {
			indices.add( tile.getDefinition() );
		}
		return indices;
	}

	public static List<Pair<String, Long>> dataToIndices( List<AnnotationData<?>> data ) {
		List<Pair<String, Long>> indices = new ArrayList<>();
		for ( AnnotationData<?> d : data ) {
			indices.add( d.getCertificate() );
		}
		return indices;
	}


    public static JSONObject certificateToJSON( Pair<String, Long> certificate ) {

        JSONObject json = new JSONObject();
        try {
            json.put( "uuid", certificate.getFirst() );
            json.put( "timestamp", certificate.getSecond().toString() );
        } catch ( Exception e ) {
            e.printStackTrace();
        }

        return json;
    }


    static public Pair<String, Long> getCertificateFromJSON( JSONObject json ) throws IllegalArgumentException {

        try {

            UUID uuid = UUID.fromString( json.getString("uuid") );
            Long timestamp = Long.parseLong( json.getString("timestamp") );
            return new Pair<>( uuid.toString(), timestamp );

        } catch ( Exception e ) {
            throw new IllegalArgumentException( e );
        }

    }


    public static AnnotationBin getBinFromJSON( JSONObject json ) throws IllegalArgumentException {

        try {

            Map<String, List<Pair<String, Long>>> certificates =  new LinkedHashMap<>();

            Iterator<?> groups = json.keys();
            while( groups.hasNext() ){

                String group = (String)groups.next();

                if( json.get(group) instanceof JSONArray ) {

                    JSONArray jsonCertificates = json.getJSONArray(group);

                    List<Pair<String, Long>> certificateList = new LinkedList<>();
                    for (int i=0; i<jsonCertificates.length(); i++) {

                        JSONObject jsonRef = jsonCertificates.getJSONObject( i );
                        certificateList.add( getCertificateFromJSON( jsonRef ) );
                    }
                    certificates.put( group, certificateList );
                }

            }
            return new AnnotationBin( certificates );

        } catch ( Exception e ) {
            throw new IllegalArgumentException( e );
        }

    }


    static public JSONObject binToJSON( AnnotationBin bin ) {

        JSONObject binJSON = new JSONObject();
        try {

            // for each group group in a bin
            for (Map.Entry<String, List<Pair<String, Long>>> certificateEntry : bin.getData().entrySet() ) {

                String group = certificateEntry.getKey();
                List<Pair<String, Long>> certificates = certificateEntry.getValue();

                JSONArray certificateJSON = new JSONArray();
                for ( Pair<String, Long> certificate : certificates ) {
                    certificateJSON.put( certificateToJSON( certificate ) );
                }

                // add group to bin json object
                binJSON.put( group, certificateJSON );
            }

        } catch ( Exception e ) {
            e.printStackTrace();
        }

        return binJSON;
    }


    static public AnnotationTile getTileFromJSON( JSONObject json ) throws IllegalArgumentException {

        try {

            TileIndex index = new TileIndex( json.getInt("level"),
                    json.getInt("x"),
                    json.getInt("y"),
                    AnnotationIndexer.NUM_BINS,
                    AnnotationIndexer.NUM_BINS );

            // create tile with empty bins
            DenseTileData<AnnotationBin> tile = new DenseTileData<>( index );

            // for all binkeys
            Iterator<?> binKeys = json.keys();
            while( binKeys.hasNext() ) {

                String binKey = (String)binKeys.next();

                if( json.get(binKey) instanceof JSONObject ){

                    JSONObject bin = (JSONObject)json.get(binKey);
                    BinIndex binIndex = BinIndex.fromString( binKey );
                    tile.setBin( binIndex.getX(), binIndex.getY(), getBinFromJSON( bin ));
                }
            }

            return new AnnotationTile( tile.getDefinition(), tile.getData() );

        } catch ( Exception e ) {
            throw new IllegalArgumentException( e );
        }

    }


    static public JSONObject tileToJSON( AnnotationTile tile ) {

        JSONObject tileJSON = new JSONObject();

        try {

            tileJSON.put("level", tile.getDefinition().getLevel() );
            tileJSON.put("x", tile.getDefinition().getX() );
            tileJSON.put("y", tile.getDefinition().getY() );

            for (int i=0; i<tile.getDefinition().getXBins(); i++ ) {
                for (int j=0; j<tile.getDefinition().getYBins(); j++ ) {

                    AnnotationBin bin = tile.getBin(i, j);

                    if ( bin != null) {
                        // add bin object to tile
                        tileJSON.put( new BinIndex(i, j).toString(), binToJSON( bin ) );
                    }

                }
            }

        } catch ( Exception e ) {
            e.printStackTrace();
        }

        return tileJSON;

    }

	
}
